[primTorch] Refs for pdist, triu and related ops

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82819

This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
~~In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.~~

Commit:
[primTorch] Added refs for pairwise_distance and pdist

[ghstack-poisoned]

Lezcano commented on torch/nn/modules/distance.py:
```diff
@@ -7,13 +7,26 @@
+
+    Returns either a scalar (if input tensors are rank 1) or a vector of a vector of all pairs of distances
+    (if input tensors are rank 2). In the second case, the output components are distances of pairs
+    ordered in the same way as hte as output of
```
the*

Commit:
Update on "[primTorch] Added refs for pairwise_distance and pdist"




[ghstack-poisoned]

Commit:
Update on "[primTorch] Added refs for pairwise_distance and pdist"




[ghstack-poisoned]

Commit:
Update on "[primTorch] Added refs for pairwise_distance and pdist"




[ghstack-poisoned]

fdrocha commented:
Note that `triu_indices`, `tril_indices` and `pdist` refs all use `torch.masked_select`
which does not currently have a ref implementation. I don't think there is a good way
of implementing them with existing refs.

I thought of also implementing `masked_select` ref but it is not clear to me how to do it.
In the tracking sheet it appears as a ref, but it seems to me
that to be implemented efficiently it either needs to be a prim or to use some other prim
that allows for indexing with boolean Tensors (which currently does not seem to exist).

On an unrelated note, I had to xfail some MetaTensor tests because `masked_select`
in general produces tensors with data-dependent shapes but in these cases the shapes
are actually known in advance. Not sure if there is some way of providing a ref override
for the meta backend that manually specifies these shapes.

fdrocha commented:
@ngimel could you PTAL?

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.

[ghstack-poisoned]

ngimel commented:
So, as you say, `masked_select`, even if we had a ref for it, would produce data-dependent shapes, and for all these ops you would know the shapes in advance. This indicates that using `masked_select` is not a correct approach to decomposing, and for eager implementations it was probably picked due to ease of implementation. Try to think of some arange-based+transformation approaches for triu/tril_indices

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -3631,6 +3635,88 @@ def trace(self: TensorLikeType) -> TensorLikeType:
+        a.ndim >= 2, lambda: "triu: input tensor must have at least 2 dimensions"
+    )
+    h, w = a.shape[-2:]
+    return a * (
```
this will result in incorrect nan propagation

fdrocha replied:
Thanks, will change it to where.

ngimel commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -577,3 +579,30 @@ def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
+def pdist(a: TensorLikeType, p: int = 2) -> TensorLikeType:
+    check(a.ndim == 2, lambda: f"pdist only supports 2D tensors, got: {a.ndim}D")
+    check(p >= 0, lambda: "pdist only supports non-negative p values")
+    t = torch.linalg.vector_norm(a.unsqueeze(1) - a, ord=p, dim=2)
```
this decomposition is quite catastrophic without compiler (even more so than usual decompositions), increasing memory use by orders of maginute, but even with compiler it's unclear if compiler would be able to recover matmul that's typically used for efficient implementation of cdist/pdist

fdrocha replied:
Hadn't thought of that, that is indeed terrible.

For `p=2` you can write this using a `matmul` with the transpose like you say (and maybe it is worth it to have that in as a special case), but for generic p I don't think you can.

I am not seeing a way of making this more efficient for generic p. You need to compute

$$ D_{ij} = \sum_a \left| a_{ia} - a_{ja}\right|^p $$

and for $p=2$ you can drop the absolute value and expand to write it in terms of `a @ a.T` but for generic `p` I can't see a way of doing without creating the intermediary tensor $M_{ija}=a_{ia} - a_{ja}$ or doing some explicit for loops. Any suggestions?

ngimel replied:
YEs, you are right, matmul solution is applicable to p=2 only (but also, let's face it, 99% of pdist calls in the wild are with p=2). It's fine to have a special case for p=2, and use your decomposition for everything else (because that's right, there's no obvious better way).

fdrocha replied:
I added this. One thought: should I document somewhere that this is expected to be inefficient for `p!=2`? I will add a comment in the code, but wondering if there is some other place to note that in.

Lezcano replied:
I think it's fine really, as this is expected if you put some thought into it.

fdrocha commented:
> So, as you say, `masked_select`, even if we had a ref for it, would produce data-dependent shapes, and for all these ops you would know the shapes in advance. This indicates that using `masked_select` is not a correct approach to decomposing, and for eager implementations it was probably picked due to ease of implementation. Try to think of some arange-based+transformation approaches for triu/tril_indices

Creating a matrix of the row or column indices is trivial with arange (the proposed ref does that), but flattening just the upper/lower triangular part of that into a vector is trickier.
I spent some time thinking about this, and the only options I see to do this in python are: 
  - Using `masked_select` or some other boolean indexing approach
  - Explicit `O(min(row,col))` python loop

Maybe there is a trick with integer rounding that will let you do it without either of these, but I haven't figured out how and it starts to feel too clever by half.

ngimel commented on torch/nn/modules/distance.py:
```diff
@@ -7,13 +7,26 @@
-        \Vert x \Vert _p = \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}.
+        \Vert \mathbf{x} \Vert _p = \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}.
+
+    Returns either a scalar (if input tensors are rank 1) or a vector of a vector of all pairs of distances
```
for tensors of rank 1 or 0 it returns 0d tensor, not scalar. Also, what is `vector of a vector`?  It's a 1d tensor with distances. It also works for tensors with a rank higher than 2

fdrocha replied:
You are right, I'll fix that. I was thinking of `pdist` when I wrote that, but this is for `pairwise_distance`.

Lezcano commented:
Stackoverflow to the rescue!
https://stackoverflow.com/questions/27086195/linear-index-upper-triangular-matrix

fdrocha commented:
> Stackoverflow to the rescue! https://stackoverflow.com/questions/27086195/linear-index-upper-triangular-matrix

Nice! I had a vague memory there should be some trick involving `sqrt` and integer rounding, that is it! It's inverse triangular numbers!
The stacoverflow answer is only for `offset=1` but shouldn't be too hard to generalize it, I will have ago.

ngimel commented:
Right, eager implementation is doing something like that I believe, but given that they needed some fixes to imprecise floor div for large numbers, I now wonder if it's really worth it https://github.com/pytorch/pytorch/pull/15203/files#diff-e98d6aff1cd0c0815422d73c81b61a8b12904e0e2b98ddc8981b1d6551cf70e7R118 (or we could limit it for the sizes where we can use sqrt directly).

Lezcano commented:
I think checking how large we can go without making mistakes with this trick in float precision and then doing something like using 64 bits and then erroring out would be the best way to go (although it'll take some effort)

fdrocha commented:
Maybe I am missing something and I will check more carefully, but I want to say that if you use 64-bit doubles, then it should work correctly until at least $n=2^{52}$ (where `n=row*col`), above which you can't store all integers in a double without loss of precision. Until this point `sqrt` is guaranteed to give you the correct integer part which is all you care about.

But if you are calling `tril_indices` with arguments that large you are going to need something like 140 petabytes store the result ( $2^{52} \times 8 \times 2$ ) - I don't think we need to worry about it for a couple of years.

Thanks Natalia for highlighting that cuda implementation! I had only checked CPU version but that cuda version spells out exactly what I need to do.

Lezcano commented:
True about doubles. The thing is that doubles on CUDA are slowwww. 
Now, it may be just fine for this computation, as it may very well be memory bound. Probably the wisest take would be just to folloe the CUDA implementation, as you guys suggested.

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.

[ghstack-poisoned]

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.

[ghstack-poisoned]

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.

[ghstack-poisoned]

Lezcano commented on torch/_refs/__init__.py:
```diff
@@ -3689,63 +3702,140 @@ def tril(a: TensorLikeType, diagonal: int = 0) -> TensorLikeType:
 
-@register_decomposition(torch.ops.aten.tril_indices)
-def tril_indices(
+def _get_triu_sizes(row: int, col: int, offset: int) -> Tuple[int, int, int]:
```
I haven't thought too much about it, but intuitively, isn't `triu_sizes` the same as `tril_sizes` but swapping the row and the column?

Lezcano replied:
Nevermind, I see that you use `tril_sizes` tin the implementation here.

Lezcano replied:
Actually, yeah, can't this function be implemented as:
```python
return _get_tril_sizes(col, row, -offset)
```
?

fdrocha replied:
No, unfortunately it doesn't work that way.

The total size (trapezoid+rectangle) is the same, but the split into trapezoid and rectangle is different.
What happens is that you can split the shape of the lower (upper) indices into a trapezoid and either a top (bottom) rectangle or a left (right) rectangle. For the purposes of the implementation you need the size of the top/bottom rectangle and when you flip rows and columns like that you map bottom rectangle into right rectangle, not the top rectangle that you need.

E.g.,:
```
> torch._refs._get_triu_sizes(2,3,-1)
(3, 3, 3)
> torch._refs._get_tril_sizes(3,2,1)
(2, 4, 2)
```

Lezcano replied:
I knew I was missing something. Could you add comments in these two functions (or one of them and reference it from the other) documenting what is it that they do exactly? Thank you!

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16196,10 +16219,15 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
-               DecorateInfo(unittest.expectedFailure, 'TestFakeTensorNonErroring', 'test_fake_autocast'),
-               DecorateInfo(unittest.expectedFailure, 'TestFakeTensorNonErroring', 'test_fake'),
-               DecorateInfo(unittest.expectedFailure, 'TestProxyTensorOpInfo', 'test_make_fx_fake_exhaustive'),
+               # test_dtypes will find this works for float dtypes, but it doesn't really
+               # make sense to use this with floats
+               DecorateInfo(unittest.skip('Skipped!'), 'TestCommon', 'test_dtypes'),
```
Then should this function support a non-integral dtype? Perhaps we should add a check on the ATen implementation... WDYT @ngimel?

fdrocha replied:
The implementations do *mostly* work for float indices, and maybe there is a use case for having it and I am just not seeing it.

To be honest, the reason I restricted to ints there is that there are some test failures for large sizes that only happen for float16/bfloat16 and it felt like it didn't really make sense to support that.

ngimel replied:
Yeah agree with @lezcano, this function should naturally support only integral dtypes (and only reasonably wide integral dtypes at that, not uint8, which is just asking for errors)

fdrocha replied:
OK, cool, I will make those changes then.

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16196,10 +16219,15 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+               # test_dtypes will find this works for float dtypes, but it doesn't really
+               # make sense to use this with floats
+               DecorateInfo(unittest.skip('Skipped!'), 'TestCommon', 'test_dtypes'),
+               # for some small dtypes, the torch cpu implementation gives the wrong results
+               # when you run it for large input, so we skip those tests
+               DecorateInfo(unittest.skip('Skipped!'), 'TestDecomp', 'test_comprehensive',
+                            device_type="cpu", dtypes=(torch.uint8, torch.int8, torch.int16)),
+               DecorateInfo(unittest.skip('Skipped!'), 'TestDecomp', 'test_quick',
+                            device_type="cpu", dtypes=(torch.uint8, torch.int8, torch.int16)),
```
Could you open an issue noting this?

fdrocha replied:
To be clear, it only gives the wrong results when you exceed the size of the dtype. I.e., if you ask for something with indices that would exceed 255 for uint8 like `torch.triu_indices(300,300, dtype=torch.uint8)`.
In those cases both current cuda implementation and the new ref implementation just give you the right result in int64 cast to the appropriate dtype, while the cpu implementation will give you 0 for the cases where the result would be over 255.

So, not clear if we should care about this. You could argue you should get an error when you exceed the dtype's max size.
Do you still think I should open an issue?

fdrocha replied:
FTR, discussed it Mario and I will make a separate PR so that CPU implementation errors when this issue arrises.

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -19070,10 +19094,8 @@ def __init__(
         supports_nvfuser=False,
         skips=(
-            # RunTimError: no _refs support for torch.Tensor.masked_select
+            # RunTimeError: no _refs support for torch.Tensor.index_select
```
We can live with this one for now. But could you post in the primtorch channel how should this one be implemented (no need to implement it in this PR of course).

fdrocha replied:
Will do. FWIW, this also requires matmul which does also not have a ref implementation yet.

Lezcano commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -577,3 +579,34 @@ def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
+    check(a.ndim == 2, lambda: f"pdist only supports 2D tensors, got: {a.ndim}D")
+    check(p >= 0, lambda: "pdist only supports non-negative p values")
+    if p == 2:
+        aTa = torch.matmul(a, a.T)
```
You may want to use `torch.mm` here as these are just 2-dimensional tensors.

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.

[ghstack-poisoned]

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
~~In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.~~

[ghstack-poisoned]

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
~~In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.~~

[ghstack-poisoned]

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
~~In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.~~

[ghstack-poisoned]

fdrocha commented:
Ok, `tri[ul]_indices` now only accept int dtypes. This should be ready, @ngimel could you take a look?

Lezcano commented on torch/nn/modules/distance.py:
```diff
@@ -7,13 +7,21 @@
 
     .. math ::
-        \Vert x \Vert _p = \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}.
+        \mathrm{dist}\left(\mathbf{v}, \mathbf{u}\right) = \left\Vert \mathbf{v}-\mathbf{u} + \epsilon \mathbf{e} \right\Vert_p,
+
+    where :math:`\mathbf{e}` is the vector of ones and the ``p``-norm is given by.
+
+    .. math ::
+        \Vert \mathbf{x} \Vert _p = \left( \sum_{i=1}^n  \vert x_i \vert ^ p \right) ^ {1/p}.
```
Why `mathbf`? Is this physics notation for vectors? We don't use that notation in PyTorch really.

fdrocha replied:
I guess it is a physics thing, i.e., bold symbols are vectors.
What do is used for vectors then? `\vec{x}`, or just `x`? I tried finding an example in the docs, but it is hard to search for and I couldn't find anything.

Lezcano replied:
We just use `x`. For example, see https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html

In general, in the linalg docs, I used capital letters for matrices (usually `A, B`) and small letters for vectors (usually `x`, `y`).

lezcano reviewed (commented):
I had a controversial thought. 
We could implement `triu` as `torch.tril(a.mT, -diagonal).mT.contiguous()`. I'll let @ngimel assess whether we want to do this.

I left two minor comments but this is pretty much good to go. Thank you for tacking these rather challenging ops @fdrocha!

fdrocha commented:
> I had a controversial thought. We could implement `triu` as `torch.tril(a.mT, -diagonal).mT.contiguous()`. I'll let @ngimel assess whether we want to do this.

Oh, nice, I like it but will wait for @ngimel to weigh in.

ngimel commented:
>We could implement triu as torch.tril(a.mT, -diagonal).mT.contiguous(). I'll let @ngimel assess whether we want to do this.

This is cool, but I think this unnecessarily complicates things for compilers, leaving 2 separate functions is fine.

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
~~In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.~~

[ghstack-poisoned]

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -3920,6 +3924,178 @@ def rop(
+    check(
+        layout == torch.strided,
+        lambda: f"only support layout=torch.strided, got {layout}",
+    )
```
should you check that dtype is int or int32?

Lezcano replied:
Probably also in the ATen function actually.

fdrocha replied:
Good point! Added the check.
@Lezcano - not sure what you mean about ATen function - you already get an error when you use other dtypes because we changed the AT_DISPATCH macro.

ngimel replied:
ATen will through an error based on the macro, that's fine.

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -3920,6 +3924,178 @@ def rop(
+    device: DeviceLikeType = "cpu",
+    pin_memory: bool = False,
+) -> TensorLikeType:
+    if pin_memory:
```
nit: you can put these check into utility fn shared between tril/triu

ngimel commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -577,3 +579,36 @@ def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
+    if p == 2:
+        aTa = torch.mm(a, a.T)
+        aTa_diag = torch.diag(aTa)
+        t = torch.sqrt(aTa_diag + aTa_diag.unsqueeze(-1) - 2 * aTa)
```
This implementation is prone to having negative values due to fp rounding errors, so typically we clamp values before doing sqrt

fdrocha replied:
Thanks, done.

ngimel commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6738,13 +6738,47 @@ def sample_inputs_tril_triu(op_info, device, dtype, requires_grad, **kwargs):
+def sample_inputs_trilu_indices(op_info, device, dtype, requires_grad, **kwargs):
+    # (row, col, offset)
+    args_list = ((1, 1),
+                 (3, 3, 1),
```
I don't think you need all these inputs as sample inputs (a lot of tests are run on sample inputs), it's better to have "interesting" small inputs as sample, and everything else as reference

fdrocha replied:
Reasonable, cut those down.

On a related note, prior to me adding the OpInfos for `triu_indices/tril_indices` they were already being tested by some tests in `test/test_cuda.py` and `test/test_tensor_creation_ops.py`, with some shared code in `common_methods_invocations.py`. I haven't looked at those very closely, but they are probably redundant with having the OpInfo based tests - do you think we should consider dropping them?
Since we are going to need OpInfos for all the tensor creation ops for PrimTorch, maybe we will want to drop `test/test_tensor_creation_ops.py` entirely at some point...

Otherwise, I made all the changes you suggested, let me know if there's anything else.

ngimel replied:
Yes, removing tests in test_tensor_creation_ops and test_cuda sounds great! (Doesn't need to be in this PR).

fdrocha replied:
Ok, will do it in a separate PR.

ngimel reviewed (commented):
I left minor comments, overall looks great!

Commit:
Update on "[primTorch] Refs for pdist, triu and related ops"


This PR adds refs for the following ops:
   - `torch.triu`
   - `torch.tril`
   - `torch.triu_indices`
   - `torch.tril_indices`
   - `torch.nn.functional.pairwise_distance`
   - `torch.nn.functional.pdist`

It adds OpInfos for
   - `torch.triu_indices`
   - `torch.tril_indices`

Note that these were already tested in `test/test_tensor_creation_ops.py`
but for the ref tests we need the OpInfos.

Finally, it improves documentation for PairwiseDistance and adds
a missing import to `torch/testing/_internal/opinfo/core.py`.

This started with an attempt to just add the `nn.functional` refs above,
but it turned out that `pdist` was easiest to implement using `triu_indices`
so I added that one and the related functions as well.
~~In the end, I changed the `pdist` implementation to not use `triu_indices`
but kept the other refs.~~

[ghstack-poisoned]

fdrocha commented on torch/_refs/__init__.py:
```diff
@@ -3920,6 +3924,178 @@ def rop(
+        - torch.arange(h, device=a.device).unsqueeze(-1)
+    ) <= diagonal
+
+    if a.dtype is torch.bool:
```
Added this as `mask_tensor` and also used it the other ref this pattern was being used in.

ngimel reviewed (approved):
Approved, please land when CI is green

fdrocha commented:
@pytorchbot merge -g
